Gifski/util.swift

```diff
@@ -295,9 +299,63 @@ extension NSAlert {
 
 
 extension AVAssetImageGenerator {
-	func generateCGImagesAsynchronously(forTimePoints timePoints: [CMTime], completionHandler: @escaping AVAssetImageGeneratorCompletionHandler) {
+	struct CompletionHandlerResult {
+		let image: CGImage?
+		let requestedTime: CMTime
+		let actualTime: CMTime
+		let completedCount: Int
+		let totalCount: Int
+		let isCancelled: Bool
+		let isFinished: Bool
+	}
+
+	/// TODO: Remove this when using Swift 5
+	enum Error: Swift.Error {}
+
+	func generateCGImagesAsynchronously(
+		forTimePoints timePoints: [CMTime],
+		completionHandler: @escaping (CoreResult<CompletionHandlerResult, Error>) -> Void
+	) {
 		let times = timePoints.map { NSValue(time: $0) }
-		generateCGImagesAsynchronously(forTimes: times, completionHandler: completionHandler)
+		let totalCount = times.count
+		var completedCount = 0
+
+		generateCGImagesAsynchronously(forTimes: times) { requestedTime, image, actualTime, result, error in
+			switch result {
+			case .succeeded:
+				completedCount += 1
+
+				completionHandler(
+					.success(
+						CompletionHandlerResult(
+							image: image,
+							requestedTime: requestedTime,
+							actualTime: actualTime,
+							completedCount: completedCount,
+							totalCount: totalCount,
+							isCancelled: false,
+							isFinished: completedCount == totalCount
+						)
+					)
+				)
+			case .failed:
+				completionHandler(.failure(error! as! Error))
+			case .cancelled:
+				completionHandler(
+					.success(
+						CompletionHandlerResult(
+							image: image,
+							requestedTime: requestedTime,
+							actualTime: actualTime,
+							completedCount: completedCount,
+							totalCount: totalCount,
+							isCancelled: true,
+							isFinished: completedCount == totalCount
+						)
+					)
+				)
```

Comment:
I know I could DRY this up, but didn't want to spend too much time on it until I decide on whether to make it a success or failure.